domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt

```diff
@@ -161,101 +190,291 @@ class TopicListController @Inject constructor(
+        ),
+        numberOfDaysPassed
+      )
+    }
+    return Pair(null, numberOfDaysPassed)
+  }
+
+  private fun createOngoingStoryListBasedOnMostRecentlyCompleted(
+    storyId: String,
+    story: StorySummary,
+    mostRecentCompletedChapterProgress: ChapterProgress,
+    completedChapterProgressList: List<ChapterProgress>,
+    topic: Topic,
+    isTopicConsideredCompleted: Boolean
+  ): Pair<PromotedStory?, Long> {
+    var numberOfDaysPassed = 0L
+    val lastChapterSummary: ChapterSummary? =
+      story.chapterList.find { chapterSummary ->
+        mostRecentCompletedChapterProgress.explorationId == chapterSummary.explorationId
+      }
+    val nextChapterIndex = story.chapterList.indexOf(lastChapterSummary) + 1
+    if (story.chapterList.size > nextChapterIndex) {
+      val nextChapterSummary: ChapterSummary? = story.chapterList[nextChapterIndex]
+      if (nextChapterSummary != null) {
+        numberOfDaysPassed = mostRecentCompletedChapterProgress.getNumberOfDaysPassed()
+        return Pair(
+          addPromotedStoryInRecommendedStoryList(
+            storyId,
+            topic,
+            completedChapterProgressList.size,
+            story.chapterCount,
+            nextChapterSummary.name,
+            nextChapterSummary.explorationId,
+            isTopicConsideredCompleted
+          ),
+          numberOfDaysPassed
+        )
+      }
     }
-    return ongoingStoryListBuilder.build()
+    return Pair(null, numberOfDaysPassed)
   }
 
-  private fun createRecommendedStoryList(): List<PromotedStory> {
-    val recommendedStories = ArrayList<PromotedStory>()
+  private fun addPromotedStoryInRecommendedStoryList(
+    storyId: String,
+    topic: Topic,
+    completedChapterProgressListSize: Int,
+    chapterCount: Int,
+    chapterSummaryName: String,
+    chapterSummaryExplorationId: String,
+    isTopicConsideredCompleted: Boolean
+  ): PromotedStory {
+    return createPromotedStory(
+      storyId,
+      topic,
+      completedChapterProgressListSize,
+      chapterCount,
+      chapterSummaryName,
+      chapterSummaryExplorationId,
+      isTopicConsideredCompleted
+    )
+  }
+
+  private fun ChapterProgress.getNumberOfDaysPassed(): Long {
+    return TimeUnit.MILLISECONDS.toDays(
+      oppiaClock.getCurrentCalendar().timeInMillis - this.lastPlayedTimestamp
+    )
+  }
+
+  private fun computeSuggestedStories(
+    topicProgressList: List<TopicProgress>
+  ): List<PromotedStory> {
+    val recommendedStories = mutableListOf<PromotedStory>()
+
     val topicIdJsonArray = jsonAssetRetriever
       .loadJsonFromAsset("topics.json")!!
       .getJSONArray("topic_id_list")
-    for (i in 0 until topicIdJsonArray.length()) {
-      createRecommendedStoryFromAssets(topicIdJsonArray[i].toString())?.let {
-        recommendedStories.add(it)
+
+    val topicIdList = (0 until topicIdJsonArray.length()).map { topicIdJsonArray[it].toString() }
+
+    val index = topicIdList.indexOf(topicProgressList.last().topicId)
```

Comment:
As discussed: I missed this earlier, but this doesn't seem to support the graph-based dependency structure the design document outlines [here](https://docs.google.com/document/d/1o7hOz1rP-hJaKn1f5x6LV6ms0bzhyCpsTifJQY5crCc/edit?ts=5f9fc81b#heading=h.990cprr9g1j5). This seems to be assuming that we have a linear pathway of topics not a graph where 1 topic may require multiple preceding topics to be completed before being recommended. We need to build the dependency graph into code & ensure that it's tested.

You can also incorporate the test topics (which will probably be needed for testing some of the more advanced cases), but please make sure to design the graph in such a way where topics can be missing (e.g. if a topic is missing then we just wouldn't recommend that lesson). In this way, you can probably build a separate part of the graph for testing purposes, e.g.:
- Test topic 0 depends on fractions
- Test topic 1 depends on test topic 0 + ratios

Or something similar. In production the test topics won't be present so they should never be recommended.

Reply:
Consider above e.g.

- Test topic 0 depends on fractions

- Test topic 1 depends on test topic 0 + ratios

If I play  `Test topic 0`  and next I play` Test topic 1` so recommended stories will be Fraction and Ratios? or based on recently played topic I have to recommend Ratios only?

Reply:
If you play test topic 0 fractions should not be recommended since you completed something that requires it. If you finish fractions, topic 0 should be recommended. Similarly, with no progress fractions should be recommended since it has no prerequisites.

Reply:
Updated PTAL.